{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This script takes the 10X output from Experiment 1 and Experiment 2 and converts the 10X output into count matrices."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Import Packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd     # Package to deal with data frames\n",
    "import scipy.io         # Used to read in the matrix.mtx file using 'scipy.io.mmread'\n",
    "import os               # Sets working directory\n",
    "import csv              # Used to read the barcodes.tsv and features.tsv files using list comprehensions\n",
    "import gzip             # Unzips .gz files from 10X output\n",
    "import numpy as np      # Used for mathematical functions\n",
    "import anndata          # Scanpy pipeline\n",
    "import scanpy as sc     # Scanpy pipeline"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Set the Current Working Directory"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%config Completer.use_jedi = False\n",
    "\n",
    "os.chdir('mydirectory')\n",
    "os.getcwd()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# All functions are loaded from a file called 'Functions.py'\n",
    "%run scripts/Functions.py"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Define Samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "samples = ['C_I_1', 'C_I_2', 'C_I_3', 'F_I_4', 'F_I_5', 'F_I_6',\n",
    "           'C_P_7', 'C_P_8', 'F_P_10', 'F_P_11', 'F_P_12']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Build Scanpy/AnnData objects"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Working on sample C_I_1\n",
      "AnnData object with n_obs × n_vars = 2125 × 32301\n",
      "    var: 'gene_ids', 'feature_types'\n",
      "Finished with sample C_I_1!\n",
      "\n",
      "\n",
      "Working on sample C_I_2\n",
      "AnnData object with n_obs × n_vars = 8111 × 32302\n",
      "    var: 'gene_ids', 'feature_types'\n",
      "Finished with sample C_I_2!\n",
      "\n",
      "\n",
      "Working on sample C_I_3\n",
      "AnnData object with n_obs × n_vars = 7528 × 32302\n",
      "    var: 'gene_ids', 'feature_types'\n",
      "Finished with sample C_I_3!\n",
      "\n",
      "\n",
      "Working on sample F_I_4\n",
      "AnnData object with n_obs × n_vars = 2990 × 32301\n",
      "    var: 'gene_ids', 'feature_types'\n",
      "Finished with sample F_I_4!\n",
      "\n",
      "\n",
      "Working on sample F_I_5\n",
      "AnnData object with n_obs × n_vars = 5018 × 32301\n",
      "    var: 'gene_ids', 'feature_types'\n",
      "Finished with sample F_I_5!\n",
      "\n",
      "\n",
      "Working on sample F_I_6\n",
      "AnnData object with n_obs × n_vars = 4302 × 32301\n",
      "    var: 'gene_ids', 'feature_types'\n",
      "Finished with sample F_I_6!\n",
      "\n",
      "\n",
      "Working on sample C_P_7\n",
      "AnnData object with n_obs × n_vars = 1435 × 32301\n",
      "    var: 'gene_ids', 'feature_types'\n",
      "Finished with sample C_P_7!\n",
      "\n",
      "\n",
      "Working on sample C_P_8\n",
      "AnnData object with n_obs × n_vars = 2681 × 32301\n",
      "    var: 'gene_ids', 'feature_types'\n",
      "Finished with sample C_P_8!\n",
      "\n",
      "\n",
      "Working on sample F_P_10\n",
      "AnnData object with n_obs × n_vars = 2412 × 32301\n",
      "    var: 'gene_ids', 'feature_types'\n",
      "Finished with sample F_P_10!\n",
      "\n",
      "\n",
      "Working on sample F_P_11\n",
      "AnnData object with n_obs × n_vars = 1443 × 32301\n",
      "    var: 'gene_ids', 'feature_types'\n",
      "Finished with sample F_P_11!\n",
      "\n",
      "\n",
      "Working on sample F_P_12\n",
      "AnnData object with n_obs × n_vars = 1605 × 32301\n",
      "    var: 'gene_ids', 'feature_types'\n",
      "Finished with sample F_P_12!\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Initialize empty list to fill in later\n",
    "Adatas = [0] * len(samples)\n",
    "\n",
    "# Loop over each sample\n",
    "for i,sample in enumerate(samples):\n",
    "    \n",
    "    print('\\nWorking on sample {}'.format(sample))\n",
    "    \n",
    "    # Get the path to the 10X data\n",
    "    data_path = 'data/CellRanger/{}/filtered_feature_bc_matrix'.format(sample)\n",
    "    \n",
    "    # Load the data as an AnnData object\n",
    "    adata = sc.read_10x_mtx(data_path, gex_only = False)\n",
    "    \n",
    "    # Save the count matrix as a Numpy array\n",
    "    adata.X = adata.X.toarray()\n",
    "    adata.X = adata.X.astype(int)\n",
    "    \n",
    "    # Save to list\n",
    "    Adatas[i] = adata\n",
    "    \n",
    "    # Print the data object\n",
    "    print(Adatas[i])\n",
    "    \n",
    "    print('Finished with sample {}!\\n'.format(sample))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Add Metadata for each sample"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Trying to set attribute `.obs` of view, copying.\n",
      "Trying to set attribute `.obs` of view, copying.\n",
      "Trying to set attribute `.obs` of view, copying.\n",
      "Trying to set attribute `.obs` of view, copying.\n",
      "Trying to set attribute `.obs` of view, copying.\n",
      "Trying to set attribute `.obs` of view, copying.\n",
      "Trying to set attribute `.obs` of view, copying.\n",
      "Trying to set attribute `.obs` of view, copying.\n",
      "Trying to set attribute `.obs` of view, copying.\n",
      "Trying to set attribute `.obs` of view, copying.\n",
      "Trying to set attribute `.obs` of view, copying.\n"
     ]
    }
   ],
   "source": [
    "# The following antibodies were used for CITE seq\n",
    "CITE_all = ['OT-1', 'CD28', 'OX40', '4-1BB', 'DR3', 'NKG2D', 'PD-1', 'TIGIT', 'LAG-3', 'CTLA4', 'Tim-3',\n",
    "             'VEGFR2', 'CD39', 'Hashtag1', 'Hashtag2', 'Hashtag3', 'hB2-M']\n",
    "\n",
    "# Loop over each sample\n",
    "for i,sample in enumerate(samples):\n",
    "    \n",
    "    # Add a column name for the sample name\n",
    "    Adatas[i].obs['sample'] = sample\n",
    "    \n",
    "    # Add a column for the genotype\n",
    "    if sample[0] == 'C':\n",
    "        Adatas[i].obs['genotype'] = 'Control'\n",
    "    if sample[0] == 'F':\n",
    "        Adatas[i].obs['genotype'] = 'Formate'\n",
    "        \n",
    "    # Add column for antibody\n",
    "    if sample.split('_')[1] == 'I':\n",
    "        Adatas[i].obs['antibody'] = 'Rat IgG2a'\n",
    "    if sample.split('_')[1] == 'P':\n",
    "        Adatas[i].obs['antibody'] = 'anti-PD-1'\n",
    "        \n",
    "    # Deal with CITE-seq data\n",
    "    CITE_cols = IntersectionOfGenes(CITE_all, Adatas[i])\n",
    "    \n",
    "    # Extract the CITE-seq columns as a pandas DataFrame\n",
    "    CITE = pd.DataFrame(index = Adatas[i].obs_names,\n",
    "                        columns = CITE_cols,\n",
    "                        data = Adatas[i][:, CITE_cols].X.astype(int))\n",
    "\n",
    "    # Add the CITE-seq data to the metadata\n",
    "    Adatas[i].obs = pd.concat([Adatas[i].obs, CITE], axis = 1)\n",
    "\n",
    "    # Remove the columns from the count matrices\n",
    "    genes_to_keep = [gene for gene in Adatas[i].var_names if gene not in CITE_cols]\n",
    "    Adatas[i] = Adatas[i][:, genes_to_keep]\n",
    "        \n",
    "    # Add total number of counts for each cell\n",
    "    Adatas[i].obs['nCounts'] = np.sum(Adatas[i].X, axis = 1).astype(int)\n",
    "    \n",
    "    # Add number of unique genes that are expressed for each cell\n",
    "    Adatas[i].obs['nGenes'] = np.count_nonzero(Adatas[i].X, axis = 1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Add TCR Metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C_I_1\n",
      "C_I_2\n",
      "C_I_3\n",
      "F_I_4\n",
      "F_I_5\n",
      "F_I_6\n",
      "C_P_7\n",
      "C_P_8\n",
      "F_P_10\n",
      "F_P_11\n",
      "F_P_12\n"
     ]
    }
   ],
   "source": [
    "# Loop over each sample\n",
    "for i,sample in enumerate(samples):\n",
    "    \n",
    "    print(sample)\n",
    "    \n",
    "    # Load the TCR data from the CreateTCRData.py pipeline\n",
    "    TCR_data = 'data/CellRanger/{}/{}_TCR.csv'.format(sample, sample)\n",
    "    TCR_data = pd.read_csv(TCR_data, index_col = 0)\n",
    "    \n",
    "    # Add the TCR data to the metadata\n",
    "    Adatas[i].obs = pd.concat([Adatas[i].obs, TCR_data], axis = 1)\n",
    "    \n",
    "    # Save TCR and AB_status as categorical data\n",
    "    for col in ['TCR', 'AB_status']:\n",
    "        Adatas[i].obs[col] = Adatas[i].obs[col].astype('category')\n",
    "        \n",
    "    # Save clone size as integer value\n",
    "    Adatas[i].obs['CloneSize_pre_filter'] = Adatas[i].obs['CloneSize_pre_filter'].astype(int)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Print out number of cells for GEX data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "For sample C_I_1, there are 2125 cells and 32285 genes.\n",
      "For sample C_I_2, there are 8111 cells and 32285 genes.\n",
      "For sample C_I_3, there are 7528 cells and 32285 genes.\n",
      "For sample F_I_4, there are 2990 cells and 32285 genes.\n",
      "For sample F_I_5, there are 5018 cells and 32285 genes.\n",
      "For sample F_I_6, there are 4302 cells and 32285 genes.\n",
      "For sample C_P_7, there are 1435 cells and 32285 genes.\n",
      "For sample C_P_8, there are 2681 cells and 32285 genes.\n",
      "For sample F_P_10, there are 2412 cells and 32285 genes.\n",
      "For sample F_P_11, there are 1443 cells and 32285 genes.\n",
      "For sample F_P_12, there are 1605 cells and 32285 genes.\n"
     ]
    }
   ],
   "source": [
    "for i,sample in enumerate(samples):\n",
    "    rows = Adatas[i].shape[0]\n",
    "    col = Adatas[i].shape[1]\n",
    "    \n",
    "    print('For sample {}, there are {} cells and {} genes.'.format(sample, rows, col))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Print out number of cells for TCR data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "For sample C_I_1, there are 580 cells with a sequenced TCR.\n",
      "For sample C_I_2, there are 5631 cells with a sequenced TCR.\n",
      "For sample C_I_3, there are 3598 cells with a sequenced TCR.\n",
      "For sample F_I_4, there are 1505 cells with a sequenced TCR.\n",
      "For sample F_I_5, there are 3224 cells with a sequenced TCR.\n",
      "For sample F_I_6, there are 2858 cells with a sequenced TCR.\n",
      "For sample C_P_7, there are 729 cells with a sequenced TCR.\n",
      "For sample C_P_8, there are 2084 cells with a sequenced TCR.\n",
      "For sample F_P_10, there are 1025 cells with a sequenced TCR.\n",
      "For sample F_P_11, there are 697 cells with a sequenced TCR.\n",
      "For sample F_P_12, there are 1091 cells with a sequenced TCR.\n"
     ]
    }
   ],
   "source": [
    "for i,sample in enumerate(samples):\n",
    "\n",
    "    # Count number of cells that has a TCR sequenced\n",
    "    rows = len(Adatas[i][ Adatas[i].obs['AB_status'] == 'AB_cell' ])\n",
    "    \n",
    "    print('For sample {}, there are {} cells with a sequenced TCR.'.format(sample, rows))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Change Cell Barcodes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# This step is done as a way to prepare for data integration later on\n",
    "for i,sample in enumerate(samples):\n",
    "    \n",
    "    # Split the old cell barcodes by the '-' character\n",
    "    # Take the nucleotide part of the barcodes (by using [0])\n",
    "    # Then add _i for each sample\n",
    "    s_number = Adatas[i].obs['sample'].unique()[0].split('_')[2]\n",
    "    Adatas[i].obs.rename(index = lambda x: x.split('-')[0] + '_' + str(s_number), inplace = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Concatenate into Single Adata Object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>sample_num</th>\n",
       "      <th>genotype</th>\n",
       "      <th>antibody</th>\n",
       "      <th>TCR</th>\n",
       "      <th>AB_status</th>\n",
       "      <th>OT-1</th>\n",
       "      <th>CD28</th>\n",
       "      <th>OX40</th>\n",
       "      <th>4-1BB</th>\n",
       "      <th>...</th>\n",
       "      <th>Tim-3</th>\n",
       "      <th>VEGFR2</th>\n",
       "      <th>CD39</th>\n",
       "      <th>Hashtag1</th>\n",
       "      <th>Hashtag2</th>\n",
       "      <th>Hashtag3</th>\n",
       "      <th>hB2-M</th>\n",
       "      <th>nCounts</th>\n",
       "      <th>nGenes</th>\n",
       "      <th>CloneSize_pre_filter</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Barcode</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACCTGCAAAGTGCG_1</th>\n",
       "      <td>C_I_1</td>\n",
       "      <td>1</td>\n",
       "      <td>Control</td>\n",
       "      <td>Rat IgG2a</td>\n",
       "      <td>notcr</td>\n",
       "      <td>notcr</td>\n",
       "      <td>10</td>\n",
       "      <td>310</td>\n",
       "      <td>608</td>\n",
       "      <td>221</td>\n",
       "      <td>...</td>\n",
       "      <td>138</td>\n",
       "      <td>77</td>\n",
       "      <td>302</td>\n",
       "      <td>6143</td>\n",
       "      <td>27</td>\n",
       "      <td>35</td>\n",
       "      <td>0</td>\n",
       "      <td>8060</td>\n",
       "      <td>2563</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGCACGAAAGC_1</th>\n",
       "      <td>C_I_1</td>\n",
       "      <td>1</td>\n",
       "      <td>Control</td>\n",
       "      <td>Rat IgG2a</td>\n",
       "      <td>notcr</td>\n",
       "      <td>notcr</td>\n",
       "      <td>2</td>\n",
       "      <td>12</td>\n",
       "      <td>21</td>\n",
       "      <td>40</td>\n",
       "      <td>...</td>\n",
       "      <td>27</td>\n",
       "      <td>0</td>\n",
       "      <td>55</td>\n",
       "      <td>1289</td>\n",
       "      <td>19</td>\n",
       "      <td>18</td>\n",
       "      <td>0</td>\n",
       "      <td>663</td>\n",
       "      <td>196</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGGTCACACGC_1</th>\n",
       "      <td>C_I_1</td>\n",
       "      <td>1</td>\n",
       "      <td>Control</td>\n",
       "      <td>Rat IgG2a</td>\n",
       "      <td>notcr</td>\n",
       "      <td>notcr</td>\n",
       "      <td>1</td>\n",
       "      <td>59</td>\n",
       "      <td>54</td>\n",
       "      <td>26</td>\n",
       "      <td>...</td>\n",
       "      <td>159</td>\n",
       "      <td>6</td>\n",
       "      <td>276</td>\n",
       "      <td>4286</td>\n",
       "      <td>32</td>\n",
       "      <td>28</td>\n",
       "      <td>0</td>\n",
       "      <td>774</td>\n",
       "      <td>299</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGTCCCTCAGT_1</th>\n",
       "      <td>C_I_1</td>\n",
       "      <td>1</td>\n",
       "      <td>Control</td>\n",
       "      <td>Rat IgG2a</td>\n",
       "      <td>CAASGTSSGQKLVF|CAWSLGVANTGQLYF</td>\n",
       "      <td>AB_cell</td>\n",
       "      <td>0</td>\n",
       "      <td>22</td>\n",
       "      <td>50</td>\n",
       "      <td>14</td>\n",
       "      <td>...</td>\n",
       "      <td>44</td>\n",
       "      <td>2</td>\n",
       "      <td>54</td>\n",
       "      <td>8</td>\n",
       "      <td>19</td>\n",
       "      <td>1473</td>\n",
       "      <td>0</td>\n",
       "      <td>9853</td>\n",
       "      <td>2977</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGTCGACCAGC_1</th>\n",
       "      <td>C_I_1</td>\n",
       "      <td>1</td>\n",
       "      <td>Control</td>\n",
       "      <td>Rat IgG2a</td>\n",
       "      <td>|CASSPRTGGFSGNTLYF</td>\n",
       "      <td>beta_chain_only</td>\n",
       "      <td>1</td>\n",
       "      <td>11</td>\n",
       "      <td>8</td>\n",
       "      <td>5</td>\n",
       "      <td>...</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>9</td>\n",
       "      <td>8</td>\n",
       "      <td>30</td>\n",
       "      <td>1442</td>\n",
       "      <td>0</td>\n",
       "      <td>4004</td>\n",
       "      <td>1421</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGCGCTCGCGTTTC_12</th>\n",
       "      <td>F_P_12</td>\n",
       "      <td>12</td>\n",
       "      <td>Formate</td>\n",
       "      <td>anti-PD-1</td>\n",
       "      <td>CALSDHPGYQNFYF|CASSGQGAREQYF</td>\n",
       "      <td>AB_cell</td>\n",
       "      <td>18</td>\n",
       "      <td>34</td>\n",
       "      <td>15</td>\n",
       "      <td>66</td>\n",
       "      <td>...</td>\n",
       "      <td>59</td>\n",
       "      <td>7</td>\n",
       "      <td>227</td>\n",
       "      <td>8</td>\n",
       "      <td>6167</td>\n",
       "      <td>14</td>\n",
       "      <td>0</td>\n",
       "      <td>18219</td>\n",
       "      <td>3829</td>\n",
       "      <td>30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGGTTCATTGGTAC_12</th>\n",
       "      <td>F_P_12</td>\n",
       "      <td>12</td>\n",
       "      <td>Formate</td>\n",
       "      <td>anti-PD-1</td>\n",
       "      <td>CAAKDYSNNRLTL|CASSRGTGEDTQYF</td>\n",
       "      <td>AB_cell</td>\n",
       "      <td>0</td>\n",
       "      <td>69</td>\n",
       "      <td>14</td>\n",
       "      <td>133</td>\n",
       "      <td>...</td>\n",
       "      <td>151</td>\n",
       "      <td>1</td>\n",
       "      <td>204</td>\n",
       "      <td>3509</td>\n",
       "      <td>86</td>\n",
       "      <td>37</td>\n",
       "      <td>0</td>\n",
       "      <td>35219</td>\n",
       "      <td>5903</td>\n",
       "      <td>26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGGTTGTCCAACTA_12</th>\n",
       "      <td>F_P_12</td>\n",
       "      <td>12</td>\n",
       "      <td>Formate</td>\n",
       "      <td>anti-PD-1</td>\n",
       "      <td>CAMREGGDNYAQGLTF|CASSLLGGAETLYF-CASSLSPGTGGTETLYF</td>\n",
       "      <td>AB_cell</td>\n",
       "      <td>0</td>\n",
       "      <td>19</td>\n",
       "      <td>73</td>\n",
       "      <td>61</td>\n",
       "      <td>...</td>\n",
       "      <td>108</td>\n",
       "      <td>0</td>\n",
       "      <td>468</td>\n",
       "      <td>2768</td>\n",
       "      <td>28</td>\n",
       "      <td>32</td>\n",
       "      <td>0</td>\n",
       "      <td>19696</td>\n",
       "      <td>4480</td>\n",
       "      <td>92</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCAGTGGTAACG_12</th>\n",
       "      <td>F_P_12</td>\n",
       "      <td>12</td>\n",
       "      <td>Formate</td>\n",
       "      <td>anti-PD-1</td>\n",
       "      <td>notcr</td>\n",
       "      <td>notcr</td>\n",
       "      <td>2</td>\n",
       "      <td>71</td>\n",
       "      <td>300</td>\n",
       "      <td>145</td>\n",
       "      <td>...</td>\n",
       "      <td>249</td>\n",
       "      <td>9</td>\n",
       "      <td>877</td>\n",
       "      <td>6776</td>\n",
       "      <td>20</td>\n",
       "      <td>26</td>\n",
       "      <td>0</td>\n",
       "      <td>20797</td>\n",
       "      <td>4540</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCATCAACACAC_12</th>\n",
       "      <td>F_P_12</td>\n",
       "      <td>12</td>\n",
       "      <td>Formate</td>\n",
       "      <td>anti-PD-1</td>\n",
       "      <td>CAMREDYGSSGNKLIF|CASSLSPGTGGTETLYF</td>\n",
       "      <td>AB_cell</td>\n",
       "      <td>4</td>\n",
       "      <td>35</td>\n",
       "      <td>22</td>\n",
       "      <td>75</td>\n",
       "      <td>...</td>\n",
       "      <td>102</td>\n",
       "      <td>3</td>\n",
       "      <td>499</td>\n",
       "      <td>3286</td>\n",
       "      <td>39</td>\n",
       "      <td>25</td>\n",
       "      <td>0</td>\n",
       "      <td>26466</td>\n",
       "      <td>5349</td>\n",
       "      <td>54</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>39650 rows × 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     sample  sample_num genotype   antibody  \\\n",
       "Barcode                                                       \n",
       "AAACCTGCAAAGTGCG_1    C_I_1           1  Control  Rat IgG2a   \n",
       "AAACCTGCACGAAAGC_1    C_I_1           1  Control  Rat IgG2a   \n",
       "AAACCTGGTCACACGC_1    C_I_1           1  Control  Rat IgG2a   \n",
       "AAACCTGTCCCTCAGT_1    C_I_1           1  Control  Rat IgG2a   \n",
       "AAACCTGTCGACCAGC_1    C_I_1           1  Control  Rat IgG2a   \n",
       "...                     ...         ...      ...        ...   \n",
       "TTTGCGCTCGCGTTTC_12  F_P_12          12  Formate  anti-PD-1   \n",
       "TTTGGTTCATTGGTAC_12  F_P_12          12  Formate  anti-PD-1   \n",
       "TTTGGTTGTCCAACTA_12  F_P_12          12  Formate  anti-PD-1   \n",
       "TTTGTCAGTGGTAACG_12  F_P_12          12  Formate  anti-PD-1   \n",
       "TTTGTCATCAACACAC_12  F_P_12          12  Formate  anti-PD-1   \n",
       "\n",
       "                                                                   TCR  \\\n",
       "Barcode                                                                  \n",
       "AAACCTGCAAAGTGCG_1                                               notcr   \n",
       "AAACCTGCACGAAAGC_1                                               notcr   \n",
       "AAACCTGGTCACACGC_1                                               notcr   \n",
       "AAACCTGTCCCTCAGT_1                      CAASGTSSGQKLVF|CAWSLGVANTGQLYF   \n",
       "AAACCTGTCGACCAGC_1                                  |CASSPRTGGFSGNTLYF   \n",
       "...                                                                ...   \n",
       "TTTGCGCTCGCGTTTC_12                       CALSDHPGYQNFYF|CASSGQGAREQYF   \n",
       "TTTGGTTCATTGGTAC_12                       CAAKDYSNNRLTL|CASSRGTGEDTQYF   \n",
       "TTTGGTTGTCCAACTA_12  CAMREGGDNYAQGLTF|CASSLLGGAETLYF-CASSLSPGTGGTETLYF   \n",
       "TTTGTCAGTGGTAACG_12                                              notcr   \n",
       "TTTGTCATCAACACAC_12                 CAMREDYGSSGNKLIF|CASSLSPGTGGTETLYF   \n",
       "\n",
       "                           AB_status  OT-1  CD28  OX40  4-1BB  ...  Tim-3  \\\n",
       "Barcode                                                        ...          \n",
       "AAACCTGCAAAGTGCG_1             notcr    10   310   608    221  ...    138   \n",
       "AAACCTGCACGAAAGC_1             notcr     2    12    21     40  ...     27   \n",
       "AAACCTGGTCACACGC_1             notcr     1    59    54     26  ...    159   \n",
       "AAACCTGTCCCTCAGT_1           AB_cell     0    22    50     14  ...     44   \n",
       "AAACCTGTCGACCAGC_1   beta_chain_only     1    11     8      5  ...      5   \n",
       "...                              ...   ...   ...   ...    ...  ...    ...   \n",
       "TTTGCGCTCGCGTTTC_12          AB_cell    18    34    15     66  ...     59   \n",
       "TTTGGTTCATTGGTAC_12          AB_cell     0    69    14    133  ...    151   \n",
       "TTTGGTTGTCCAACTA_12          AB_cell     0    19    73     61  ...    108   \n",
       "TTTGTCAGTGGTAACG_12            notcr     2    71   300    145  ...    249   \n",
       "TTTGTCATCAACACAC_12          AB_cell     4    35    22     75  ...    102   \n",
       "\n",
       "                     VEGFR2  CD39  Hashtag1  Hashtag2  Hashtag3  hB2-M  \\\n",
       "Barcode                                                                  \n",
       "AAACCTGCAAAGTGCG_1       77   302      6143        27        35      0   \n",
       "AAACCTGCACGAAAGC_1        0    55      1289        19        18      0   \n",
       "AAACCTGGTCACACGC_1        6   276      4286        32        28      0   \n",
       "AAACCTGTCCCTCAGT_1        2    54         8        19      1473      0   \n",
       "AAACCTGTCGACCAGC_1        2     9         8        30      1442      0   \n",
       "...                     ...   ...       ...       ...       ...    ...   \n",
       "TTTGCGCTCGCGTTTC_12       7   227         8      6167        14      0   \n",
       "TTTGGTTCATTGGTAC_12       1   204      3509        86        37      0   \n",
       "TTTGGTTGTCCAACTA_12       0   468      2768        28        32      0   \n",
       "TTTGTCAGTGGTAACG_12       9   877      6776        20        26      0   \n",
       "TTTGTCATCAACACAC_12       3   499      3286        39        25      0   \n",
       "\n",
       "                     nCounts  nGenes  CloneSize_pre_filter  \n",
       "Barcode                                                     \n",
       "AAACCTGCAAAGTGCG_1      8060    2563                     0  \n",
       "AAACCTGCACGAAAGC_1       663     196                     0  \n",
       "AAACCTGGTCACACGC_1       774     299                     0  \n",
       "AAACCTGTCCCTCAGT_1      9853    2977                    20  \n",
       "AAACCTGTCGACCAGC_1      4004    1421                     0  \n",
       "...                      ...     ...                   ...  \n",
       "TTTGCGCTCGCGTTTC_12    18219    3829                    30  \n",
       "TTTGGTTCATTGGTAC_12    35219    5903                    26  \n",
       "TTTGGTTGTCCAACTA_12    19696    4480                    92  \n",
       "TTTGTCAGTGGTAACG_12    20797    4540                     0  \n",
       "TTTGTCATCAACACAC_12    26466    5349                    54  \n",
       "\n",
       "[39650 rows x 26 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Create a Python generator\n",
    "generator = (Adatas[i] for i in range(1, len(Adatas)))\n",
    "\n",
    "# Concatenate into a single object\n",
    "Adata = Adatas[0].concatenate(generator, index_unique = None)\n",
    "\n",
    "# Rename the 'batch' column to 'sample_num'\n",
    "Adata.obs = Adata.obs.rename(columns = {'batch':'sample_num'})\n",
    "\n",
    "# Renumber the sample_num column\n",
    "Adata.obs['sample_num'] = [int(x.split('_')[2]) for x in Adata.obs['sample']]\n",
    "\n",
    "# Reorder columns\n",
    "# Order by GEX/sample/TCR data\n",
    "cols = ['sample', 'sample_num', 'genotype', 'antibody', 'TCR', 'AB_status', 'nCounts', 'nGenes', 'CloneSize_pre_filter'  ]\n",
    "\n",
    "Adata.obs = Adata.obs[cols]\n",
    "\n",
    "# Replace NaN values\n",
    "for col in ['TCR', 'AB_status']:\n",
    "    Adata.obs[col] = Adata.obs[col].fillna('notcr')\n",
    "    \n",
    "for col in ['CloneSize_pre_filter']:\n",
    "    Adata.obs[col] = Adata.obs[col].fillna(0).astype(int)\n",
    "\n",
    "Adata.obs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Save File"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "... storing 'sample' as categorical\n",
      "... storing 'genotype' as categorical\n",
      "... storing 'antibody' as categorical\n",
      "... storing 'TCR' as categorical\n",
      "... storing 'feature_types' as categorical\n"
     ]
    }
   ],
   "source": [
    "Adata.write_h5ad('data/ScanpyObjects/MergedData_prefilter.h5ad')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Jared",
   "language": "python",
   "name": "jared"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
